packages/server/src/internals/procedure.ts

```diff
@@ -97,22 +86,26 @@ function getParseFn<TParsedInput>(
 /**
  * @internal
  */
-export class Procedure<TInputContext, TContext, TInput, TParsedInput, TOutput> {
+export class Procedure<TInputContext, TContext, TInput, TOutput> {
   private middlewares: Readonly<Array<MiddlewareFunction<any, any>>>;
-  private resolver: ProcedureResolver<TContext, TParsedInput, TOutput>;
-  private readonly inputParser: ProcedureInputParser<TParsedInput>;
-  private parse: ParseFn<TParsedInput>;
+  private resolver: ProcedureResolver<TContext, TInput, TOutput>;
+  public inputParser: ProcedureParser<TInput>;
+  private parseInputFn: ParseFn<TInput>;
+  public outputParser: ProcedureParser<TOutput> | undefined;
+  private parseOutputFn: ParseFn<TOutput> | undefined;
 
-  constructor(opts: ProcedureOptions<TContext, TParsedInput, TOutput>) {
+  constructor(opts: ProcedureOptions<TContext, TInput, TOutput>) {
     this.middlewares = opts.middlewares;
     this.resolver = opts.resolver;
     this.inputParser = opts.inputParser;
-    this.parse = getParseFn(this.inputParser);
+    this.parseInputFn = getParseFn(this.inputParser);
+    this.outputParser = opts.outputParser;
+    this.parseOutputFn = this.outputParser && getParseFn(this.outputParser);
```

Comment:
This could maybe be simplified for usage it just has a passthrough fn

```suggestion
    this.parseOutputFn = this.outputParser ?? (val) => val;
```

Reply:
outdated